### DISCUSSION TITLE
Monads for the Rest of Us

### DISCUSSION THREAD

**Discussion by arialdomartini** [arialdomartini/arialdomartini.github.io#26] [General]
Discussions about the article series [Monads for the Rest of Us](https://arialdomartini.github.io/monads-for-the-rest-of-us)

**Comment by freynder** [comment]
Great article. I especially like the fact you briefly mention category theory to illustrate the core definition vs implementations / interpretations / visualizations. What also became more clear was that languages that support function types can actually demonstrate the function lifting interpretation for `fmap` and `bind` , not just in Haskell. Thank you!

**Comment by alex1yaremchuk** [comment]
I don't like that examples are not easily runnable.  
Especially intermediary ones.
That's a bit frustrating learning such a subtle topic.

**Reply by arialdomartini** [reply]
Hi @alex1yaremchuk, 
thank you for the feedback. Would you please elaborate? Do you think, for example, that having a clonable repository with a C# solution would help? Is so, I will be happy to add it!

**Reply by alex1yaremchuk** [reply]
Thank you for a prompt and kind reply, sorry if my first message was a bit harsh.

I just having a bit hard time with getting back to C#, it's overall not being very friendly after TypeScript.

What I'd like to have is "batteries included" and runnable attitude. 
I think running examples is big part of learning.
In this sense I'd like it use Console.WriteLine instead of Log.Information or the instructions to how to install the needed package.
I think Assert is not much needed here as well, but as it is here I'd like to have a brief instruction on how to install xUnit or other package. 

There is a typo I believe - f takes B as a param but is applied to A.
static class FunctionExtensions
{
    internal static Func<A, C> ComposedWith<A, B, C>(this Func<B, C> f, Func<A, B> g) => a => g(f(a));
}

Overall I've conquered troubles and moving on. 
I'll let you know the essential result of my learning later - at the moment I'm at part 4.

**Reply by arialdomartini** [reply]
Not harsh at all! On the contrary, you are giving me very valuable suggestions.
The idea of a battery-included experience is cool. 

>  In this sense I'd like it use Console.WriteLine instead of Log.Information 

Here I am not sure what you mean...

> or the instructions to how to install the needed package.
> I think Assert is not much needed here as well, but as it is here I'd like to have a brief instruction on 
> how to install xUnit or other package.

The more I think about it, the more I believe providing a separate repository with a working solution would be of help.
I will keep you informed. I would love if you could be my beta-tester!

> There is a typo I believe - f takes B as a param but is applied to A.

Yes! It should be `f(g(a))` instead of `g(f(a))`. Thank you for finding it!

> I'll let you know the essential result of my learning later - at the moment I'm at part 4.

Cool. Feel free to reach me out should you need any help or should you have more hints.

**Reply by alex1yaremchuk** [reply]
>>Cool. Feel free to reach me out should you need any help or should you have more hints.

Thanks, I got a question on part 4 - why 'bResult.Run();' not just return bResult  (as types suggest)?

```csharp
IO<B> Apply<A, B>(this Func<A, IO<B>> f, IO<A> a) => 
    new IO<B>(() =>
    {
        A aResult = a.Run();
        IO<B> bResult = f(aResult);
        return bResult.Run();
    });

```
>>I would love if you could be my beta-tester!

Sure, I'll check it out with my pleasure.

**Reply by arialdomartini** [reply]
`bResult` is of type `IO`, and we are returning a `new IO()`

```csharp
IO Apply<A, B>(this Func<A, IO> f, IO a) =>
  new IO(() =>               // This is IO
  {
    A aResult = a.Run();
    IO bResult = f(aResult);
    return bResult;          // this would be an IO, inside another IO
  });
```

This would result in a `IO` inside another `IO`

To make the types happy, we could do remove the first `new IO(() =>`:


```csharp
IO Apply<A, B>(this Func<A, IO> f, IO a)
{
    A aResult = a.Run();
    IO bResult = f(aResult);
    return bResult;          // this is IO too
}
```

The problem, as you see, is that we would *run* `a`, together with its side effects, immediately. Composing Monads is really about deferring the execution of side effects to the last moment, when the monad is finally run.

That's why we start by returning a `new IO(() => ...` which will represent a computation without executing it.


(hints: in Markdown you can surround code words with back ticks \`, and even get syntax highlighting writing 


    ```csharp
    your code here
    ```

**Reply by alex1yaremchuk** [reply]
That's right, thank you! got it now!

It became clear as I started running this section.

**Comment by alex1yaremchuk** [comment]
Well, I've finished the course and I enjoyed it a lot. 

Thank you!

Everything is much-much clearer now.

The entertaining style of writing is just wonderful - it helped a lot!

I'll put some questions in separate messages.

**Comment by alex1yaremchuk** [comment]
How can I understand that Functor has no Return/Unit function? 
I mean if there is "f a" is it not that we can put "a" into context?

And more - all the implementations are using putting values into context (and I think it is just return/unit is it not?). 

And definition from category theory says that there is mapping for objects and for functions.

To sum it up - what's the difference between return/unit/pure and putting values into context?

**Reply by alex1yaremchuk** [reply]
Well, I understand now that "f a" is just "a value of type f". 
So definition of fmap doesn't demand of functor knowing how to put value into context. 
But still.. 
1) definition in category theory says there is a mapping for objects (values).
How to understand it - there is a mapping but we can't use it?

2) all the implementations operate with putting values into context.

3) for IO I can put same value into different contexts - just use different side-effects (output different messages) - so each side-effect makes it's own monad?

**Comment by alex1yaremchuk** [comment]
I found IO monad to be very interesting and eye-opening. 
When we need to do something with a value in IO (first value) we open it but put the whole thing in context of a new monadic value (second value). 
So first value is really opened only when we need to know second value and open it.

is this "deferred" effect intrinsic to IO or some other monads have it as well?

other examples from the course resolve immediately, right? 
so they are 'transparent boxes' if I put it correctly.

**Reply by arialdomartini** [reply]
I would say that most monads have this deferral effect. For example, the State Monad encapsulates a `s -> (s, a)` function which is actually run at the end of the chain.

But yes, you are right, there are other monads where this deferral is less apparent. For example, in `List` both `map` and `bind` immediately cycle the combinations of elements.

**Comment by alex1yaremchuk** [comment]
I'll continue collecting my thoughts on monads here - we'll be able to clean up later if needed. 

Unrelated thing from Monads in pictures. 
They give an example of partial appicaton > (+) <$> (Just 5)
And they say that result is Just (+5). 

And it looks like Haskell would confirm that. 

but it wouldn't! you need to apply function completely to get something on the screen. 

that's seems to be a pity to me. 
I'd like really execute "> (+) <$> (Just 5)" and receive "Just (+5)". 
so to speak "get the touch of it".

**Reply by arialdomartini** [reply]
Yep, this is unfortunate. As far as I know, there is no implementation of `Show` for functions, so their body cannot be printed in GHCI. All I could do is:

```haskell
ghci> f = (+) <$> (Just 5)
ghci> :info f
f :: Num a => Maybe (a -> a)    -- Defined at <interactive>:12:1
ghci> :type f
f :: Num a => Maybe (a -> a)
```

**Comment by rerdavies** [comment]
Is there a section on "that's insane, why would anyone do that?!!".  Asking for a friend. 

Ok. Asking very sincerely. That seems like an awful lot of trouble. Why would anyone want to do that?

**Reply by arialdomartini** [reply]
Yes, it is!
But with a caveat: there is, I think, a disproportionate, unbalanced difference between *defining* a state monad from the scratch (what the article is about) and *just using* it.

It's a pity that I haven't made in time to publish the Part III yet: with it I aimed to show how a State Monad is used, in practice, and how easy in fact it is.

I claim that using FP is easier than many think. Yes, getting into details and learning each building block is hard. But, maybe, not harder than understanding some internals of OOP (I think to dynamic dispatching, virtual tables, message passing, method resolution order etc).

Using monads, once they have been built, on the other hand, is easier than one might think. If you have used C#'s LINQ, you have used a list monad. In many languages offering alternatives to null, you might have used the Maybe monad, or even an Either. 

If you look to the Haskell example of chapter 12

```haskell
index (Node l r) =
    do {
        ll <- index l
        rr <- index r
        return buildNode ll rr
    }
```

I would not say that this is insane. On the contrary, this is linear just like an ordinary imperative code is, while still retaining the benefits of purity.


That said, yes: venturing in FP can be daunting. If this pays off - because of other benefits - is debatable, and you will find people ready to defend either positions.

**Reply by rerdavies** [reply]
> ... while still retaining the benefits of **_purity_**.

Which are....?

**Reply by arialdomartini** [reply]
I am not sure if this is a polemic question.

If you are interested in a general discussion about the pros and cons of functional programming, I'm afraid that none of my posts cover this, I'm sorry. The good news is, you can find very easily tons of articles, blog posts and book covering exactly that.

**Comment by rerdavies** [comment]
It's a genuine question. I've read a little about functional programming.
But I've never been able to understand why monads are good.

**Reply by arialdomartini** [reply]
That's a very tough question, and as you imagine it would deserve a separate post.
Here's my take.

In general, each style has got its pros and cons. Although in love with FP, by no means am I claiming it is the only good way to go.

There is a reason why goto is evil and structured programming is a thing: it dramatically helps governing complexity. Reading a source code based on heavy use of `goto` makes comprehension hard because basically it forces the reader to run a mental interpreter. The reader has to jump around, like the CPU would, to figure out the shape of the workflow. Balanced cycles, functions, sub-routines are way better for the human mind: they help the human to see the workflow structure. Of course, implementing structured programming comes with the need to use specific patterns, for the cases that `goto` would easily solve. In other words, structured programming (like OOP and any other discipline) comes with a cost.

A similar process happens with mutable state: an algorithm based on state mutation forces the reader to mentally run the code, keeping track of each state change. 

Interpreting an algorithms such as:

```csharp
#include <bits/stdc++.h>
using namespace std;

int partition(vector<int>& arr, int low, int high)
{
    int pivot = arr[high];
    int i = (low - 1);

    for (int j = low; j <= high - 1; j++)
    {
        if (arr[j] <= pivot)
        {
            i++;
            swap(arr[i], arr[j]);
        }
    }
    swap(arr[i + 1], arr[high]);
    return (i + 1);
}

void quickSort(vector<int>& arr, int low, int high)
{
    if (low < high)
    {
        int pi = partition(arr, low, high);

        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

void printArray(const vector<int>& arr)
{
    for (int i : arr)
        cout << i << " ";
    cout << endl;
}
```

requires the reader to follow how the value of `i`, `j`, `low`, `high`, `arr` change overtime.

The same algorithm, with immutable values:

```haskell
sort :: Ord a => [a] -> [a]
sort [] = []
sort (x:xs) =
    smallerSorted ++ [x] ++ largerSorted
    where
        smallerSorted = sort (filter (<= x) xs)
        largerSorted  = sort (filter (> x) xs)
```

is possibly more explicit in conveying the idea that Quick Sort is about 

- putting a pivot value `x` between the values smaller than it and the values larger than it, where the algorithms is applied recursively.

The level of abstraction is possibly higher, but the reader does not need to mentally run the state machine to interpret how it works.

Often, algorithms based on immutability are easier to maintain than their correspondent mutable, imperative ones, because of [Locality of Behaviour](https://htmx.org/essays/locality-of-behaviour) or [Local Reasoning](https://www.dreamsongs.com/Files/PatternsOfSoftware.pdf). In few words: when a single function exclusively depends on its input, and when its outcome is completely described by its output, that function can be completely reasoned about just reading it. When there is either dependency on a global input (an implicit input parameter), or an effect on some shared state, then that function cannot be completely reasoned about in isolation.

A function that exclusively depends on its input parameters and whose only outcome is the output value can be studied in isolation and we call it *a pure function*. It can be even replaced with its own value (i.e., it is equivalent to a dictionary).

While dealing with those pure functions simplifies the local interpretation, it forces the developer to use specific patterns to build abstractions. The mere decision to never have a shared state and to only use immutable values leads to a set of practices and patterns that need to be studied. Monads are one of them.

That said, I wouldn't say that "Monads are good". Immutability and purity are good. Monads are necessary to have immutability and purity in some circumstances. They are a necessary evil, if you will.

Monads are a means to an end. You have to use monads if you want to handle a state while being immutable. Or if you want to model a shared variable without having any shared variable. Or to model a side effect without actually having a side effect.
